app/Services/Reports/Volumes/ImageAnnotations/AbundanceReportGenerator.php

```diff
@@ -38,31 +38,66 @@ class AbundanceReportGenerator extends AnnotationReportGenerator
+            } else {
+                $trees = LabelTree::whereIn('id', $rows->keys())->pluck('name', 'id');
+            }
 
             foreach ($trees as $id => $name) {
-                $rowGroup = $rows->get($id);
-                $labels = Label::whereIn('id', $rowGroup->pluck('label_id')->unique())->get();
-                $this->tmpFiles[] = $this->createCsv($rowGroup, $name, $labels);
+                if ($this->shouldUseAllLabels()) {
+                    $labels = $allLabels->get($id);
+                } else {
+                    $labelIds = $rows->get($id)->pluck('label_id')->unique();
+                    $labels = Label::whereIn('id', $labelIds)->get();
+                }
+                $this->tmpFiles[] = $this->createCsv($rows->flatten(), $name, $labels, $emptyImagesQuery);
             }
-        } elseif ($this->shouldSeparateUsers() && $rows->isNotEmpty()) {
-            $labels = Label::whereIn('id', $rows->pluck('label_id')->unique())->get();
+        } elseif ($this->shouldSeparateUsers()) {
             $rows = $rows->groupBy('user_id');
             $users = User::whereIn('id', $rows->keys())
                 ->selectRaw("id, concat(firstname, ' ', lastname) as name")
                 ->pluck('name', 'id');
+            $labels = null;
+
+            if ($this->shouldUseAllLabels()) {
+                $labels = $this->getVolumeLabels();
+            }
 
             foreach ($users as $id => $name) {
+                // Get query for images that have no annotations
+                $emptyImagesQuery = $query->clone()->whereNull('label_id')->select('filename');
+                $usedImageFilenames = $rows->get($id)->pluck('filename')->unique();
+                $usersEmptyImagesQuery = $query->clone()->select('filename')->whereNotIn('filename', $usedImageFilenames);
+                // Add query to process unannotated images (by current user) in chunks later
+                $emptyImagesQuery->union($usersEmptyImagesQuery);
+
                 $rowGroup = $rows->get($id);
-                $this->tmpFiles[] = $this->createCsv($rowGroup, $name, $labels);
+
+                if (!$this->shouldUseAllLabels()) {
+                    $labels = Label::whereIn('id', $rowGroup->pluck('label_id'))->get();
+                }
+
+                $this->tmpFiles[] = $this->createCsv($rowGroup, $name, $labels, $emptyImagesQuery);
             }
         } else {
-            $labels = Label::whereIn('id', $rows->pluck('label_id')->unique())->get();
-            $this->tmpFiles[] = $this->createCsv($rows, $this->source->name, $labels);
+            // Get query for images that have no annotations
+            $emptyImagesQuery = $query->clone()->whereNull('label_id')->select('filename');
+            $allLabelIds = $rows->pluck('label_id')->reject(fn ($k) => !$k);
+            $labels = $this->shouldUseAllLabels() ? $this->getVolumeLabels() : Label::whereIn('id', $allLabelIds)->get();
```

Comment:
Don't collect all label IDs if you may not need them:

```suggestion
            if ($this->shouldUseAllLabels()) {
                $labels = $this->getVolumeLabels();
            } else {
                $allLabelIds = $rows->pluck('label_id')->filter()->unique();
                $labels = Label::whereIn('id', $allLabelIds)->get();
            }
```